src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs

```diff
@@ -202,14 +202,33 @@ public override long Position
         public override void Flush()
         {
             EnsureNotDisposed();
+            if (_mode == CompressionMode.Compress)
+                FlushBuffers();
         }
 
         public override Task FlushAsync(CancellationToken cancellationToken)
         {
             EnsureNotDisposed();
-            return cancellationToken.IsCancellationRequested ?
-                Task.FromCanceled(cancellationToken) :
-                Task.CompletedTask;
+            return
+                cancellationToken.IsCancellationRequested ? Task.FromCanceled(cancellationToken) :
+                _mode != CompressionMode.Compress ? Task.CompletedTask :
+                FlushAsyncCore(cancellationToken);
+        }
+
+        private async Task FlushAsyncCore(CancellationToken cancellationToken)
+        {
+            if (_asyncOperations != 0)
```

Comment:
This check should be done at the beginning of FlushAsync rather than at the beginning of FlushAsyncCore.  Where it currently is, the exception will be returned out via the Task rather than out via the synchronous call.  The invalid usage will also be hidden from the developer if one of the other cases is true, e.g. if cancellation was requested or if we're decompressing.